posts/inside-rust/2020-11-11-source-based-code-coverage.md

```diff
@@ -0,0 +1,80 @@
+---
+layout: post
+title: Source-based code coverage in nightly
+author: Tyler Mandry
+team: The Compiler Team <https://www.rust-lang.org/governance/teams/compiler>
+---
+
+
+Support has landed in the nightly compiler for source-based code coverage,
+and we want your help testing it!
+
+
+# What is _source-based_ code coverage, exactly?
+
+You may already be familiar with code coverage, which shows you which lines
+of code execute. Code coverage is usually applied to tests to find out which
+code is actually being tested and which code isn’t.
+
+Nightly Rust already supports another kind of source code coverage, commonly
+called gcov, which relies on debug info to map from LLVM IR to lines of
+source code. Instrumentation is then added in the LLVM backend during code
+generation to count how many times each line is run.
+
+However, since LLVM doesn’t know exactly how Rust code is structured, there’s
+a lot lost in the translation between Rust source and LLVM IR. Line-level
+granularity is sometimes too coarse, and debug info can be unreliable,
+especially when building in release mode. The result is coverage reports that
+only show an approximation of what code actually executed.
+
+Source-based code coverage instrumentation happens directly in the Rust
+compiler. The compiler can see branches in individual expressions within the
+same line or spanning multiple lines, and this analysis is always precise
+because it's being done in MIR. That means that things like short-circuited
+conditionals, closures, and match guards are all precisely counted. Counters
+are injected as additional MIR code statements, so the compiler can still
+optimize the program without affecting the coverage results.
+
+[![Comparison of gcov and source-based coverage results][comparison-img]][comparison-img]
```

Comment:
Are you sure the image link is right? When I "View" this file, the image shows as a broken link.

https://github.com/rust-lang/blog.rust-lang.org/blob/b0de7311f34fc75e1435bca7f38bc1235dff81ea/posts/inside-rust/2020-11-11-source-based-code-coverage.md

(I posted a similar rust-lang/rust documentation PR recently, and the image worked in that case. Not sure if this repo's structure works differently.)

Reply:
Yeah, I built the site locally to test. Other posts in this repo use the same pattern but it's not going to work on github because of the absolute path.